Adding "Why Geneva" section. (#123)
lancedb/docs: docs/geneva/why-geneva.mdx

---
title: "Geneva: Feature Engineering at Startup Speed (and Scale)"
sidebarTitle: "Why Geneva?"
description: "A practical pitch for ML engineers: turn feature code into table columns, run reliable backfills at scale, and skip the KubeRay plumbing."
icon: rocket
keywords:
  [
    "geneva",
    "feature engineering",
    "lancedb enterprise",
    "udf",
    "ray",
    "kuberay",
    "backfill",
    "materialized view",
  ]
---

<Badge color="red">Enterprise-only</Badge>

> You have a small team, a fast-moving product, and a dataset that keeps getting wider every week.

Reviewer: ```suggestion
Imagine you are part of a small team with a fast-moving product, and a dataset that keeps getting wider every week.
```

One day you're shipping a recommendation tweak, the next you're computing image embeddings,
then you decide you need OCR, then sentiment, then dedup signals, then you get the idea.

Feature engineering is supposed to be the fun part: turning raw data into signals your model can use.
In practice, it often turns into a second job: building and operating the machinery needed to compute
those features reliably and repeatedly.

Geneva, the feature engineering layer in LanceDB Enterprise, is written from
the perspective of an ML engineer who wants to move fast *without*
becoming the on-call owner of a bespoke distributed system.

## TL;DR

Geneva helps you:

- **Define features as columns** using Python UDFs (`Table.add_columns()`).
- **Backfill those columns at scale** as a checkpointed, resumable background job (`Table.backfill()`).
- **Scale from laptop to Kubernetes** using Ray under the hood, while abstracting away the Ray/KubeRay details.
- **Keep iteration tight** with filters, incremental refresh, and materialized views for training layouts.
- **Stay sane operationally** with dependency packaging, retries, conflict handling, and a job console.

> If you want the short version: Geneva makes “write Python, run it everywhere, and store the result next to the data” at scale feel like a normal database operation.

Reviewer: ```suggestion
> Geneva makes “write Python functions, run it at scale, and store the result next to the source data” feel like a normal database operation.
```

## The Startup Feature Engineering Tax

> At a startup, the constraints are usually clear:

Reviewer: ```suggestion
As an AI engineer, the constraints are usually clear:
```

- You need new signals quickly (the model quality curve is unforgiving).
- Your data is increasingly multimodal (text, images, audio, video, embeddings, metadata).
- You do not have a dedicated platform team to build and maintain a feature pipeline stack.

> The first prototype is easy. The second is still fun. The tenth is where things go off the rails:

Reviewer: ```suggestion
As you iterate on your prototypes, more and more problems become apparent:
```

- How do you **recompute a feature** when your UDF changes?
- How do you **resume** when a long backfill dies halfway through?
- How do you **run expensive model inference** across millions of rows without reinventing a job system?
- How do you avoid “works on my laptop” when the cluster is Linux and you're on macOS?
- How do you keep this from becoming an Airflow + Ray + Docker + Kubernetes side project?

> Geneva’s bet is simple: feature engineering should look like **table operations**.

Reviewer: ```suggestion
Geneva’s premise is simple: feature engineering should look like **table operations**.
```

## “But My Feature Code Has Dependencies…”

This is where many teams get stuck: the notebook has `pillow`, `torch`, and a handful of “one more thing” packages.
The cluster has a different Python version, or Linux wheels, or missing environment variables.

By default, Geneva will **package your local environment** (including `site-packages` and your working directory) and ship it to workers.
When you need more control, you can define a reusable **manifest**:

<CodeGroup>
```python Python icon="python"
from geneva.manifest.builder import GenevaManifestBuilder

manifest_name = "prod-features"
manifest = (
    GenevaManifestBuilder()
    .name(manifest_name)
    .skip_site_packages(False)
    .pip(["lancedb", "numpy", "pillow"])
    .build()
)

db.define_manifest(manifest_name, manifest)

with db.context(cluster="geneva-features", manifest=manifest_name):
    tbl.backfill("title_len")
```
</CodeGroup>

And if you ever hit the dreaded “it fails on the workers but not locally” problem, Geneva includes a dependency verification tool
to compare local and Ray worker environments.

## Make Long Jobs Boring

> The fastest way to lose feature velocity is when backfills become scary.

Reviewer: ```suggestion
Nothing kills feature velocity faster than backfills that are brittle, or operationally risky.
```
